Source/DafnyLanguageServer.Test/Synchronization/CachingTest.cs

```diff
@@ -337,6 +353,85 @@ import Library
     await AssertNoDiagnosticsAreComing(CancellationToken);
   }
 
+  [Fact]
+  public async Task DocumentAddedToExistingProjectDoesNotCrash() {
+    var source1 = @"
+method Foo() {
+ var b: int := true;
+}";
+    var source2 = @"
+method Foo() {
+ var b: int := true;
+}";
+
+    var temp = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+    var file1 = CreateAndOpenTestDocument(source1, Path.Combine(temp, "source1.dfy"));
+    var file2 = CreateAndOpenTestDocument(source2, Path.Combine(temp, "source2.dfy"));
+    var project = await CreateOpenAndWaitForResolve("", Path.Combine(temp, "dfyconfig.toml"));
+    await client.WaitForNotificationCompletionAsync(project.Uri, CancellationToken);
+    ApplyChange(ref file1, new Range(0, 0, 0, 0), "// added this comment\n");
+    var diagnostics1 = await GetLastDiagnostics(file1);
+    var diagnostics2 = await GetLastDiagnostics(file2);
+    var combined = diagnostics1.Concat(diagnostics2);
+    AssertM.Equal(3, combined.Count, "diagnostics: " + string.Join("\n", combined));
+  }
+
+  [Fact]
+  public async Task ConcurrentCompilationDoesNotBreakCaching() {
+    var largeImport1 = GetLargeFile("Imported1", 100);
+    var largeImport2 = GetLargeFile("Imported2", 100);
+
+    var importerSource = @"module Importer {
+  import Imported1
+  import Imported2
+  method Bar() {
+    Imported1.Foo0();
+    Imported2.Foo0();
+  }
+}";
+
+    var temp = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+    var imported1 = CreateAndOpenTestDocument(largeImport1, Path.Combine(temp, "imported1.dfy"));
+    var imported2 = CreateAndOpenTestDocument(largeImport2, Path.Combine(temp, "imported2.dfy"));
+    var importer = CreateAndOpenTestDocument(importerSource, Path.Combine(temp, "importer.dfy"));
+    var project = CreateOpenAndWaitForResolve("", Path.Combine(temp, "dfyconfig.toml"));
+
+    var before1 = await WaitAndCountHits(imported1);
+    var before2 = await WaitAndCountHits(imported2);
+    var beforeImporter = await WaitAndCountHits(importer);
+
+    ApplyChange(ref importer, new Range(0, 0, 0, 0), "// added this comment\n");
+    for (int i = 0; i < 100; i++) {
+      ApplyChange(ref importer, new Range(0, 0, 0, 0), "// added this comment\n");
+    }
+
+    var after1 = await WaitAndCountHits(imported1);
+    var after2 = await WaitAndCountHits(imported2);
+    var afterImporter = await WaitAndCountHits(importer);
+    Assert.Equal(before1.ParseMisses, after1.ParseMisses);
+    Assert.Equal(before1.ResolutionMisses, after1.ResolutionMisses);
+    Assert.Equal(before2.ParseMisses, after2.ParseMisses);
+    Assert.Equal(before2.ResolutionMisses, after2.ResolutionMisses);
+    // Somehow, importer can have many parse hits, even though it always gets changed.
```

Comment:
I don't plan on investing time into an investigation here. Probably I should reword the comment so it's purely informational instead of also aspirational.